### PR TITLE
feat: session updates along with the event

### PR DESCRIPTION
**Pull request by marandaneto**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [X] Refactoring

## :scroll: Description
feat: session updates along with the event

## :bulb: Motivation and Context
we'd like to send a session update along with events in a single envelope

## :pencil: Checklist

- [X] I reviewed submitted code
- [X] I added tests to verify changes
- [X] All tests passing
- [X] No breaking changes

### CODE REVIEW

File: sentry-android-core/src/main/java/io/sentry/android/core/ManifestMetadataReader.java [hunk 1]
```diff
@@ -84,6 +84,7 @@ static void applyMetadata(
         options.setEnableSessionTracking(sessionTrackingEnabled);
 
         if (options.getSampleRate() == null) {
+          // TODO: it needs to read a Float I guess
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/ManifestMetadataReader.java:87]: not because of that, `ManifestMetadataReader` is using `getDouble` and it should be `getFloat` if I am not mistaken or it will read the default value

File: sentry-core/src/main/java/io/sentry/core/OutboxSender.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                && !envelope.getHeader().getEventId().equals(event.getEventId())) {
+              logger.log(
+                  SentryLevel.ERROR,
+                  "Item %d of has a different event id (%s) to the envelope header (%s)",
+                  items,
+                  envelope.getHeader().getEventId(),
+                  event.getEventId());
+              continue;
+            }
+            hub.captureEvent(event, hint);
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/OutboxSender.java:139]: this is the outbox, the events in the outbox are raw eg NDK or RN and need to be enriched, so the envelope needs to be open and events need to be processed, `captureEvent` is the way to go or the `captureEnvelope` would need to wrap all the items and wrap it as well, as the outbox only caches 1 event per envelope and no other type, it made sense to keep as it is

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java [hunk 3]
```diff
@@ -88,43 +88,83 @@ public SentryClient(final @NotNull SentryOptions options, @Nullable Connection c
...
+      final SentryEnvelope envelope = buildEnvelope(event, session);
+
+      if (envelope != null) {
+        connection.send(envelope, hint);
+      }
     } catch (IOException e) {
-      options
-          .getLogger()
-          .log(SentryLevel.WARNING, "Capturing event " + event.getEventId() + " failed.", e);
+      options.getLogger().log(SentryLevel.WARNING, "Capturing event " + sentryId + " failed.", e);
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryClient.java:131]: we could use the string args and %s

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java [hunk 4]
```diff
@@ -135,8 +175,11 @@ public SentryClient(final @NotNull SentryOptions options, @Nullable Connection c
    * @param scope the Scope or null
    */
   @TestOnly
-  void updateSessionData(
+  @Nullable
+  Session updateSessionData(
       final @NotNull SentryEvent event, final @Nullable Object hint, final @Nullable Scope scope) {
+    Session clonedSession = null;
+
     if (ApplyScopeUtils.shouldApplyScopeData(hint)) {
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryClient.java:183]: `applyScope` already checks for that

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java [hunk 5]
```diff
@@ -88,43 +88,83 @@ public SentryClient(final @NotNull SentryOptions options, @Nullable Connection c
...
-      return SentryId.EMPTY_ID;
+      if (!sample()) {
+        options
+            .getLogger()
+            .log(
+                SentryLevel.DEBUG,
+                "Event %s was dropped due to sampling decision.",
+                event.getEventId());
+        sendEvent = false;
+        event = null;
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/SentryClient.java:109]: If I'm not missing something, if we're setting this to null here, we'll continue the execution and attempt to send the event.

Shouldn't we just return SentryId.Empty here?

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java [hunk 6]
```diff
@@ -63,14 +60,17 @@ public SentryClient(final @NotNull SentryOptions options, @Nullable Connection c
...
 
+    boolean sendEvent = true;
+
     if (ApplyScopeUtils.shouldApplyScopeData(hint)) {
       // Event has already passed through here before it was cached
       // Going through again could be reading data that is no longer relevant
       // i.e proguard id, app version, threads
       event = applyScope(event, scope, hint);
 
       if (event == null) {
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/SentryClient.java:71]: Could the event be null after calling `applyScope`? 
Shouldn't we check at the start of the method, then only on beforeSend, and event processors?

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java [hunk 7]
```diff
@@ -88,43 +88,83 @@ public SentryClient(final @NotNull SentryOptions options, @Nullable Connection c
...
+                SentryLevel.DEBUG,
+                "Event %s was dropped due to sampling decision.",
+                event.getEventId());
+        sendEvent = false;
+        event = null;
+      }
     }
 
-    event = executeBeforeSend(event, hint);
+    if (event != null) {
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/SentryClient.java:113]: I find that the combination of the boolean `sendEvent` and the multiple null check is too confusing. I found it hard to follow to make sure this should work in all cases.

What about we just guard at the start for null (and return), then at the two locations where user callbacks can set it to null?

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java [hunk 8]
```diff
@@ -88,43 +88,83 @@ public SentryClient(final @NotNull SentryOptions options, @Nullable Connection c
...
 
+    SentryId sentryId = SentryId.EMPTY_ID;
+
     try {
-      connection.send(event, hint);
+      final SentryEnvelope envelope = buildEnvelope(event, session);
+
+      if (envelope != null) {
+        connection.send(envelope, hint);
+      }
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/SentryClient.java:129]: When would envelope be null here?
Should we log this out?

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java [hunk 9]
```diff
@@ -160,28 +203,22 @@ void updateSessionData(
...
-                    // otherwise we just cache in the disk but do not flush to the network.
-                    sessionHint = new SessionUpdateHint();
                   }
-                  captureSession(session, sessionHint);
                 }
               } else {
                 options.getLogger().log(SentryLevel.INFO, "Session is null on scope.withSession");
               }
             });
+        clonedSession = scope.cloneSession();
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/SentryClient.java:216]: We're cloning outside of the lock so we could be cloning a different session than the one updated above. For example if a thread closing a session was waiting on a lock to close the session and takes the lock before we call `clone` here.

Maybe we clone it inside the callback and assign the reference to a variable declared outside of the block. Ugly but this is guaranteed to clone the right state.

The method on the session (like `update`) could return the clone on each operation already to avoid having that public clone that also takes a lock

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java (hunk 5 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryClient.java:109]: the event won't be sent, the flag is false and the event is now null, the session needs to be sent though, so the flow needs to continue

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java (hunk 6 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryClient.java:71]: the event can be null when running the scope's processor

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java (hunk 7 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryClient.java:113]: i will look to get rid of one of them, but i had a reason to not do so, just dont remember now, tomorrow i look into it

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java (hunk 8 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryClient.java:129]: if the event was discarded and there is no session to be sent,  the `buildEnvelope` would return null, as there's nothing to be sent

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java (hunk 9 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryClient.java:216]: good catch, I'll need to use an `AtomicReference` maybe because it's a lambda and I cannot assign the value to not even a final object outside of the lambda or to refactor a little bit this, i will have a look

File: sentry-core/src/main/java/io/sentry/core/SentryClient.java (hunk 8 above)
Comment (bruno-garcia) [reply sentry-core/src/main/java/io/sentry/core/SentryClient.java:129]: Got it. thanks

File: sentry-core/src/main/java/io/sentry/core/Scope.java [hunk 10]
```diff
@@ -392,11 +392,19 @@ public void addEventProcessor(@NotNull EventProcessor eventProcessor) {
...
+  Session withSession(@NotNull IWithSession sessionCallback) {
+    Session cloneSession = null;
     synchronized (sessionLock) {
       sessionCallback.accept(session);
+
+      if (session != null) {
+        cloneSession = session.clone();
+      }
     }
+    return cloneSession;
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/Scope.java:407]: L: We could just return from within the synchronized block and avoid the variable
